### PR TITLE
end to end test

### PR DESCRIPTION
**Pull request by danielnorberg**
# Hey, I just made a Pull Request!

## Description
End to end tests of styx running on real GKE, datastore, bigtable, etc.

## Motivation and Context
Get automated feedback earlier in the change life-cycle instead of waiting for results from deployment to the internal staging environment.

## Checklist for PR author(s)

- [x] Changes are covered by unit test
- [x] All tests pass
- [x] Code coverage check passes
- [x] Error handling is tested
- [ ] Errors are handled at the appropriate layer
- [ ] Errors that cannot be handled where they occur are propagated
- [ ] (optional) Changes are covered by system test
- [ ] Relevant documentation updated
- [x] This PR has NO breaking change to public API
- [ ] This PR has breaking change to public API and it is documented

## Checklist for PR reviewer(s)

- [ ] This PR has been incorporated in release note for the coming version
- [x] Risky changes introduced by this PR have been all considered

### CODE REVIEW

**Comment by danielnorberg** [comment]
Not sure why codecov is unhappy. Jacoco is happy: https://6305-74431644-gh.circle-artifacts.com/0/jacoco-aggregate/styx-scheduler-service/com.spotify.styx/StyxScheduler.html

**Comment by danielnorberg** [comment]
I do not understand why the codecov patch check is not happy. From what I can tell, every line in the codecov diff is covered.

File: styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java [hunk 1]
```diff
@@ -643,7 +646,7 @@ private static ServiceAccountKeyManager createServiceAccountKeyManager() {
     }
   }
 
-  static NamespacedKubernetesClient getKubernetesClient(Config rootConfig, String id,
+  public static NamespacedKubernetesClient getKubernetesClient(Config rootConfig, String id,
```
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java]: This doesn't need to be `public`, it seems.

File: styx-e2e-test/src/test/java/com/spotify/styx/e2e_tests/DatastoreCleanupTest.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This removes old datastore test namespaces. It is not really a test.
+ */
+public class DatastoreCleanupTest {
```
Comment (honnix) [inline styx-e2e-test/src/test/java/com/spotify/styx/e2e_tests/DatastoreCleanupTest.java]: I'm a bit confused. This class together with `KubernetesCleanupTest`, when are they expected to run? Wouldn't the order matter?

Comment (honnix) [reply styx-e2e-test/src/test/java/com/spotify/styx/e2e_tests/DatastoreCleanupTest.java]: I can see in `tearDown` of base class we do try to clean up. Is this as an extra safety net? But again, this test may run in between other tests.

**Comment by honnix** [comment]
Looks good overall. Good job!

File: styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java (hunk 1 above)
Comment (danielnorberg) [reply styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java]: 👍

File: styx-e2e-test/src/test/java/com/spotify/styx/e2e_tests/DatastoreCleanupTest.java (hunk 2 above)
Comment (danielnorberg) [reply styx-e2e-test/src/test/java/com/spotify/styx/e2e_tests/DatastoreCleanupTest.java]: Added a comment.

These cleanup classes remove test accounts/namespaces etc that might be left behind by previous e2e runs that were interrupted or teardown failed.

File: styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java [hunk 3]
```diff
@@ -611,6 +610,10 @@ private static DockerRunner createDockerRunner(
         serviceAccountKeyManager, debug, styxEnvironment, secretWhitelist));
   }
 
+  public static NamespacedKubernetesClient getKubernetesClient(Config config, String id) {
```
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java]: `@VisibleForTesting`?

File: styx-e2e-test/src/test/java/com/spotify/styx/e2e_tests/EndToEndTestBase.java [hunk 4]
```diff
@@ -95,13 +97,23 @@
...
-  int styxApiPort = -1;
-
   private Datastore datastore;
   private NamespacedKubernetesClient k8s;
 
   private Config schedulerConfig;
 
+  private static final int BASE_PORT = 18080;
+
+  private final List<? extends Class<? extends EndToEndTestBase>> testClasses =
```
Comment (honnix) [inline styx-e2e-test/src/test/java/com/spotify/styx/e2e_tests/EndToEndTestBase.java]: This is smart!

File: pom.xml [hunk 5]
```diff
@@ -738,6 +765,20 @@
 
       <plugin>
         <artifactId>maven-failsafe-plugin</artifactId>
+        <version>3.0.0-M3</version>
```
Comment (honnix) [inline pom.xml]: Shall we move the version thing to pluginManagement instead? The same goes for surefire plugin.

File: styx-e2e-test/src/test/java/com/spotify/styx/e2e_tests/EndToEndTestBase.java (hunk 4 above)
Comment (danielnorberg) [reply styx-e2e-test/src/test/java/com/spotify/styx/e2e_tests/EndToEndTestBase.java]: 😬

File: pom.xml (hunk 5 above)
Comment (danielnorberg) [reply pom.xml]: 👍

File: styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java (hunk 3 above)
Comment (danielnorberg) [reply styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java]: 👍

**Pull request merged**